src/libopensc/card-piv.c

```diff
@@ -4389,12 +4384,72 @@ static int piv_get_pin_preference(sc_card_t *card, int *pin_ref)
 	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
 }
 
-static int
-piv_yk_pin_policy(sc_card_t *card, u8 *pin_policy)
+static void
+piv_yk_metadata_get_policy(sc_context_t *ctx, u8 *buf, size_t buflen, u8 *pin, u8 *touch)
 {
+	size_t policylen;
+	const u8 *policy = sc_asn1_find_tag(ctx, buf, buflen, 0x02, &policylen);
+	if (policy && policylen == 2) {
+		*pin = policy[0];
+		*touch = policy[1];
+	}
+}
+
+static void
+piv_yk_get_metadata(sc_card_t *card, u8 slot, u8 *pin_policy, u8 *touch_policy)
+{
+	sc_apdu_t apdu;
+	u8 resp[100];
+	size_t i;
 	piv_private_data_t *priv = PIV_DATA(card);
 
-	*pin_policy = priv->yk_metadata.pin.pin_policy;
+	/* initialize with the default behaviour */
+	if (pin_policy)
+		*pin_policy = 0x00;
+	if (touch_policy)
+		*pin_policy = 0x00;
+
+	if (priv->yubico_version < 0x00050300) {
+		sc_log(card->ctx, "Yubikey's PIN and touch policy not available");
+		return;
+	}
+
+	for (i = 0; i < (sizeof(priv->yk_metadata)/sizeof(*priv->yk_metadata)); i++) {
+		if (priv->yk_metadata[i].slot == 0x00)
+			/* reached the last initialized entry */
+			break;
+
+		if (priv->yk_metadata[i].slot == slot)
+			/* metadata already initialized */
+			break;
+	}
+
+	if (priv->yk_metadata[i].slot == 0x00) {
+		/* initialize this entry */
+		sc_format_apdu_ex(&apdu, 0x00, 0xF7, 0x00, slot, NULL, 0, resp, sizeof resp);
+		if (SC_SUCCESS == sc_transmit_apdu(card, &apdu) && SC_SUCCESS == sc_check_sw(card, apdu.sw1, apdu.sw2)) {
+			piv_yk_metadata_get_policy(card->ctx, resp, apdu.resplen,
+					&priv->yk_metadata[i].pin_policy,
+					&priv->yk_metadata[i].touch_policy);
+			sc_log(card->ctx, "PIN policy for slot 0x%02X: 0x%02X (touch 0x%02X)",
+					slot, priv->yk_metadata[i].pin_policy, priv->yk_metadata[i].touch_policy);
```

Comment:
I would propose to have here some return value to be able to distinguish (at least in the debug logs without parsing the ASN1 output manually), if the policy was not found or if it is set to 0 (default). Right now, the log will show up the same for these two cases.

Reply:
I added a log message